sklearn/tree/_tree.pyx

```diff
@@ -91,22 +91,29 @@ cdef class TreeBuilder:
         """Build a decision tree from the training set (X, y)."""
         pass
 
-    cdef inline _check_input(self, object X, cnp.ndarray y,
-                             cnp.ndarray sample_weight):
+    cdef inline _check_input(
+        self,
+        object X,
+        cnp.ndarray y,
+        cnp.ndarray sample_weight
+    ):
         """Check input dtype, layout and format"""
         if issparse(X):
+            # csc is fortran by default
             X = X.tocsc()
             X.sort_indices()
 
+            # if we need to copy to convert the datatype, then do it using fortran
             if X.data.dtype != DTYPE:
-                X.data = np.ascontiguousarray(X.data, dtype=DTYPE)
+                X.data = np.asfortranarray(X.data, dtype=DTYPE)
 
             if X.indices.dtype != np.int32 or X.indptr.dtype != np.int32:
                 raise ValueError("No support for np.int64 index based "
                                  "sparse matrices")
-
-        elif X.dtype != DTYPE:
+        else:
```

Comment:
I'm concerned with the memory copy here. If a large `C` ordered array is passed into `RandomForest`, then every Decision Tree will make a copy of `X`. We can mitigate this, by convert the array to "F" ordered during `RandomForest.fit`, but that is still a memory copy in the beginning.

For me, I'll want to see memory and runtime benchmarks comparing this PR and `main` with `RandomForest` to see if this memory copy is worth it. Specifically, the benchmark needs to look at large C and F ordered input.

Reply:
Good point. I will run the RandomForest asv benchmarks.